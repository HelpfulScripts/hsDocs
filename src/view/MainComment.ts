/**
 * Processing of comments.
 */

/** */
import { m, Vnode } from 'hslayout';
import { markDown } from '../markdown';
import { example }  from './MainExample';

/**
 * Main comment processing. The result appears directly below the title in the main panel.
 * Function parameters are not reported in short form here since it is assumed they will be listed 
 * individually below the main comment
 * @param mdl the module to scan for comments
 * @return a vnode representing the comment entries
 */
export function commentLong(mdl:any):Vnode { 
    let content:any[] = [];
    if (mdl.comment) {
        content.push(textOrShortTextOrDescription(mdl.comment, false));
        content.push(returns(mdl.comment, false));
        content.push(commentRemainder(mdl.comment));
    }
    return m('.hsdocs_comment', content);
}


/**
 * Report the item's description. This can come in different forms that are all handled here:
 * - comment.shortText: <description>
 * - comment.text: <description>
 * - comment.tags[{tag:'description}, text:<description>]
 * Any resulting comment will be translated from markdown to html and returned as a `Vnode`.
 * @param comment the comment object to parse
 * @param short boolean; if true, only the first paragraph of the description will be returned
 */
function textOrShortTextOrDescription(comment:any, short:boolean):Vnode {
    let text = (comment.shortText || '');
    if (comment.text) { text += '\n'+ (comment.text || ''); }
    if (comment.tags) {
        comment.tags.map((tag:any) => {
            switch (tag.tag) {
                case 'shortText':
                case 'description': text = `${text}\n${tag.text}`; break;
                case 'param': text = `${text}\n- **${tag.param}**: ${tag.text}`; break;
            }
        });
    }
    // search for pattern <example...<file...</example>
    text = text.replace(/<example[^<]*?(<file[\S\s]*?)<\/example>/gi, short? '' : example);
    return m('.hsdocs_comment_desc', prettifyCode(text, short));
}

/**
 * creates the `returns` message for a function or method.
 */
function returns(comment:any, short:boolean):Vnode {
    let text = comment.returns;
    return m('hsdocs_comment_return', !text? '': [            
        m('span.hsdocs_comment_return_tag', 'returns:'), 
        m('span.hsdocs_comment_return_text', text)
    ]);
}

function commentRemainder(comment:any):string|Vnode {
    return m('', Object.keys(comment).map((tag:any) => {
            switch(tag) {
                case 'tags':        // already handled
                case 'shortText':   // already handled
                case 'text':        // already handled
                case 'description': // already handled
                case 'returns':     // already handled
                        return '';
                default: return m('.hsdocs_comment_remainder', [
                    m('span.hsdocs_comment_return_tag', tag), 
                    m('span.hsdocs_comment_return_text', comment[tag])
                ]);
            }
    }));
}



/**
 * finds segments of `<code>...</code>` in `comment` and replaces them with a prettified version.
 * Currently the function performs operations in this sequence:
 * - add indentation for brackets {...}
 * - wrap the &lt;code&gt;...&lt;/code&gt; part within &lt;pre&gt;...&lt;/pre&gt; brackets
 * - apply markdown syntax
 * - wrap results in a trusted mithril node
 * @param comment the comment comment 
 * @param short only return the first paragraph
 */
export function prettifyCode(comment:string, short:boolean):Vnode { 
    let result = comment;

    function braceIndenting(text:string): string {
        let indent = 0;
        const result = text
            .substring(6, text.length-7)    // remove <code> and </code>
            .trim()
            .replace(/(<)/g, '&lt;').replace(/(>)/g, '&gt;')
            .split('\n')
            .map((l:string) => {
                let oldIndent = indent;
                let k = l.trim();
                indent += Math.max(-1, Math.min(1, k.split('{').length - k.split('}').length)); 
                indent += Math.max(-1, Math.min(1, k.split('[').length - k.split(']').length)); 
                return '<span class="hs-code-indent"></span>'.repeat(((indent < oldIndent)?indent:oldIndent)) + k;
            })
            .join('\n')
            .trim();
        return '<pre><code>' + result + '</code></pre>';
    }

    result = result.replace(/<code>([\S\s]*?)<\/code>/gi, braceIndenting);
    return m.trust(markDown(result, short, m.route.get()));
}
