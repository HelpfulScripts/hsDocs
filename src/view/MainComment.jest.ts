import { commentLong, comment } from './MainComment';
import { m } from 'hslayout';


const mdl = {
    comment: {
        text: `
        # complete comment
        with sub paragraphs and code:
        <code>
        line 1 {
        indented line
        }
        </code>
        `,
        shortText: `short example comment.`,
        tags: [{tag: 'description2', text:'other tag'}, {tag: 'other'}]
    },
    parameters: [{a: 5, b:'hi'}]
};

const mdl2 = {
    comment: {
        text: 'comment',
        tags: [{tag: 'description', text:'other tag'}, {tag: 'other'}],
        special: 0
    },
    parameters: [{a: 5, b:'hi'}]
};

m.route.get = () => '/api/hsDocs/0';

describe('comment', () => {
    let c1, c2, c3;

    beforeAll((done) => {
        c1 = comment(mdl);
        c2 = comment(mdl, true);
        c3 = comment(mdl2, true);
        setTimeout(done, 100);
    });

    it('should exist', ()=> {
        expect(comment).toBeDefined();
    });

    it('should translate comment, not short', (done) => {
        expect(c1).toMatchSnapshot();
        done();
    });

    it('should translate comment, short', (done) => {
        expect(c2).toMatchSnapshot();
        done();
    });

    it('should translate description comment', (done) => {
        expect(c3).toMatchSnapshot();
        done();
    });
});

describe('commentLong', () => {
    let c;

    beforeAll((done) => {
        c = commentLong(mdl);
        setTimeout(done, 100);
    });

    it('should exist', ()=> {
        expect(commentLong).toBeDefined();
    });

    it('should translate commentLong', (done) => {
        expect(c).toMatchSnapshot();
        done();
    });
});

