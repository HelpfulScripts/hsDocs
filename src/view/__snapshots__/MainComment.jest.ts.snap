// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comment should translate comment, not short 1`] = `
Object {
  "attrs": Object {
    "className": "hs-item-comment",
  },
  "children": Array [
    Object {
      "attrs": Object {
        "className": "hs-item-comment-desc",
      },
      "children": Array [
        Object {
          "attrs": undefined,
          "children": "<p>short example comment.</p>
<pre><code>    # complete comment
    with sub paragraphs and code:
    &lt;pre&gt;<code>line 1 {
&lt;span class=\\"hs-code-indent\\"&gt;&lt;/span&gt;indented line
}</code>&lt;/pre&gt;
</code></pre>",
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "<",
          "text": undefined,
        },
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": undefined,
      "key": undefined,
      "state": undefined,
      "tag": "div",
      "text": undefined,
    },
    Object {
      "attrs": null,
      "children": Array [
        Object {
          "attrs": Object {
            "className": "hs-item-comment-special",
          },
          "children": Array [
            Object {
              "attrs": Object {
                "className": "hs-item-comment-tag",
              },
              "children": undefined,
              "dom": undefined,
              "domSize": undefined,
              "events": undefined,
              "instance": undefined,
              "key": undefined,
              "state": undefined,
              "tag": "span",
              "text": "description2",
            },
            Object {
              "attrs": Object {
                "className": "hs-item-comment-text",
              },
              "children": undefined,
              "dom": undefined,
              "domSize": undefined,
              "events": undefined,
              "instance": undefined,
              "key": undefined,
              "state": undefined,
              "tag": "span",
              "text": "other tag",
            },
          ],
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "div",
          "text": undefined,
        },
        Object {
          "attrs": Object {
            "className": "hs-item-comment-special",
          },
          "children": Array [
            Object {
              "attrs": Object {
                "className": "hs-item-comment-tag",
              },
              "children": undefined,
              "dom": undefined,
              "domSize": undefined,
              "events": undefined,
              "instance": undefined,
              "key": undefined,
              "state": undefined,
              "tag": "span",
              "text": "other",
            },
            Object {
              "attrs": Object {
                "className": "hs-item-comment-text",
              },
              "children": Array [],
              "dom": undefined,
              "domSize": undefined,
              "events": undefined,
              "instance": undefined,
              "key": undefined,
              "state": undefined,
              "tag": "span",
              "text": undefined,
            },
          ],
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "div",
          "text": undefined,
        },
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": undefined,
      "key": undefined,
      "state": undefined,
      "tag": "div",
      "text": undefined,
    },
    Object {
      "attrs": Object {
        "className": "hs-item-comment-params",
      },
      "children": Array [
        Object {
          "attrs": Object {
            "className": "hs-item-comment-param",
          },
          "children": Array [
            Object {
              "attrs": Object {
                "className": "hs-item-comment-tag",
              },
              "children": undefined,
              "dom": undefined,
              "domSize": undefined,
              "events": undefined,
              "instance": undefined,
              "key": undefined,
              "state": undefined,
              "tag": "span",
              "text": "undefined:",
            },
            Object {
              "attrs": Object {
                "className": "hs-item-comment-text",
              },
              "children": undefined,
              "dom": undefined,
              "domSize": undefined,
              "events": undefined,
              "instance": undefined,
              "key": undefined,
              "state": undefined,
              "tag": "span",
              "text": "",
            },
          ],
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "div",
          "text": undefined,
        },
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": undefined,
      "key": undefined,
      "state": undefined,
      "tag": "div",
      "text": undefined,
    },
    Object {
      "attrs": Object {
        "className": "hs-item-comment-return",
      },
      "children": undefined,
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": undefined,
      "key": undefined,
      "state": undefined,
      "tag": "div",
      "text": "",
    },
    Object {
      "attrs": null,
      "children": Array [
        Object {
          "attrs": undefined,
          "children": "",
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "#",
          "text": undefined,
        },
        Object {
          "attrs": undefined,
          "children": "",
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "#",
          "text": undefined,
        },
        Object {
          "attrs": undefined,
          "children": "",
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "#",
          "text": undefined,
        },
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": undefined,
      "key": undefined,
      "state": undefined,
      "tag": "div",
      "text": undefined,
    },
  ],
  "dom": undefined,
  "domSize": undefined,
  "events": undefined,
  "instance": undefined,
  "key": undefined,
  "state": undefined,
  "tag": "div",
  "text": undefined,
}
`;

exports[`comment should translate comment, short 1`] = `
Object {
  "attrs": Object {
    "className": "hs-item-comment",
  },
  "children": Array [
    Object {
      "attrs": Object {
        "className": "hs-item-comment-desc",
      },
      "children": Array [
        Object {
          "attrs": undefined,
          "children": "<p>short example comment.",
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "<",
          "text": undefined,
        },
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": undefined,
      "key": undefined,
      "state": undefined,
      "tag": "div",
      "text": undefined,
    },
    Object {
      "attrs": Object {
        "className": "hs-item-comment-return",
      },
      "children": undefined,
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": undefined,
      "key": undefined,
      "state": undefined,
      "tag": "div",
      "text": "",
    },
    Object {
      "attrs": null,
      "children": Array [
        Object {
          "attrs": undefined,
          "children": "",
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "#",
          "text": undefined,
        },
        Object {
          "attrs": undefined,
          "children": "",
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "#",
          "text": undefined,
        },
        Object {
          "attrs": undefined,
          "children": "",
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "#",
          "text": undefined,
        },
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": undefined,
      "key": undefined,
      "state": undefined,
      "tag": "div",
      "text": undefined,
    },
  ],
  "dom": undefined,
  "domSize": undefined,
  "events": undefined,
  "instance": undefined,
  "key": undefined,
  "state": undefined,
  "tag": "div",
  "text": undefined,
}
`;

exports[`comment should translate description comment 1`] = `
Object {
  "attrs": Object {
    "className": "hs-item-comment",
  },
  "children": Array [
    Object {
      "attrs": Object {
        "className": "hs-item-comment-desc",
      },
      "children": Array [
        Object {
          "attrs": undefined,
          "children": "<p>comment
other tag",
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "<",
          "text": undefined,
        },
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": undefined,
      "key": undefined,
      "state": undefined,
      "tag": "div",
      "text": undefined,
    },
    Object {
      "attrs": Object {
        "className": "hs-item-comment-return",
      },
      "children": undefined,
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": undefined,
      "key": undefined,
      "state": undefined,
      "tag": "div",
      "text": "",
    },
    Object {
      "attrs": null,
      "children": Array [
        Object {
          "attrs": undefined,
          "children": "",
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "#",
          "text": undefined,
        },
        Object {
          "attrs": undefined,
          "children": "",
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "#",
          "text": undefined,
        },
        Object {
          "attrs": Object {
            "className": "hs-item-comment-special",
          },
          "children": Array [
            Object {
              "attrs": Object {
                "className": "hs-item-comment-tag",
              },
              "children": undefined,
              "dom": undefined,
              "domSize": undefined,
              "events": undefined,
              "instance": undefined,
              "key": undefined,
              "state": undefined,
              "tag": "span",
              "text": "special",
            },
            Object {
              "attrs": Object {
                "className": "hs-item-comment-text",
              },
              "children": undefined,
              "dom": undefined,
              "domSize": undefined,
              "events": undefined,
              "instance": undefined,
              "key": undefined,
              "state": undefined,
              "tag": "span",
              "text": "0",
            },
          ],
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "div",
          "text": undefined,
        },
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": undefined,
      "key": undefined,
      "state": undefined,
      "tag": "div",
      "text": undefined,
    },
  ],
  "dom": undefined,
  "domSize": undefined,
  "events": undefined,
  "instance": undefined,
  "key": undefined,
  "state": undefined,
  "tag": "div",
  "text": undefined,
}
`;

exports[`commentLong should translate commentLong 1`] = `
Object {
  "attrs": Object {
    "className": "hs-item-comment",
  },
  "children": Array [
    Object {
      "attrs": Object {
        "className": "hs-item-comment-desc",
      },
      "children": Array [
        Object {
          "attrs": undefined,
          "children": "<p>short example comment.</p>
<pre><code>    # complete comment
    with sub paragraphs and code:
    &lt;pre&gt;<code>line 1 {
&lt;span class=\\"hs-code-indent\\"&gt;&lt;/span&gt;indented line
}</code>&lt;/pre&gt;
</code></pre>",
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "<",
          "text": undefined,
        },
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": undefined,
      "key": undefined,
      "state": undefined,
      "tag": "div",
      "text": undefined,
    },
    Object {
      "attrs": Object {
        "className": "hs-item-comment-return",
      },
      "children": undefined,
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": undefined,
      "key": undefined,
      "state": undefined,
      "tag": "div",
      "text": "",
    },
    Object {
      "attrs": null,
      "children": Array [
        Object {
          "attrs": undefined,
          "children": "",
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "#",
          "text": undefined,
        },
        Object {
          "attrs": undefined,
          "children": "",
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "#",
          "text": undefined,
        },
        Object {
          "attrs": undefined,
          "children": "",
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "instance": undefined,
          "key": undefined,
          "state": undefined,
          "tag": "#",
          "text": undefined,
        },
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": undefined,
      "key": undefined,
      "state": undefined,
      "tag": "div",
      "text": undefined,
    },
  ],
  "dom": undefined,
  "domSize": undefined,
  "events": undefined,
  "instance": undefined,
  "key": undefined,
  "state": undefined,
  "tag": "div",
  "text": undefined,
}
`;
